flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/generated/CompileUtils.java

```diff
@@ -45,7 +48,7 @@
 	 * number of Meta zone GC (class unloading), resulting in performance bottlenecks. So we add
 	 * a cache to avoid this problem.
 	 */
-	protected static final Cache<Tuple2<ClassLoader, String>, Class<?>> COMPILED_CACHE = CacheBuilder
+	protected static final Cache<String, Cache<ClassLoader, Class>> COMPILED_CACHE = CacheBuilder
 		.newBuilder()
 		.maximumSize(100)   // estimated cache size
```

Comment:
Why not invoke `.softValues()` here instead of a new `Cache` value ?

Reply:
My hunch is that using soft values in the inner cache gives us fine grained garbage collection. Instead of removing the whole inner cache, we may want to only remove the least recently used class instances.
According to this [reading](http://jeremymanson.blogspot.com/2009/07/how-hotspot-decides-to-clear_07.html), the timestamp that a soft value was lastly accessed plays a part when JVM decides whether to clear the reference.

Reply:
Do you mean there are multiple code values for one class name ? How could that happen ?

Reply:
The origin cache uses tuple of class loader and class name as key. So yes, I guess theoretically there can be multiple class instances for the same name, otherwise we don't need to involve class loaders at all. Perhaps it can happen if we run multiple queries concurrently in one TM. @JingsongLi any insights about this?

Reply:
First, different tasks of one Job maybe have different class loader, so yes, same code maybe have multiple Java Classes.
Second, same sql run in different client maybe generate same codes too.

Reply:
Why couldn't we just cache the code not the class ?

Reply:
Why cache code? For what?
Can you see above comments in `CompileUtils.COMPILED_CACHE`, we cache classes for performance and GC.

Reply:
Sorry, i didn't have much knowledge, i'm just asking my questions. I also noticed that Apache Spark[1] and Apache Calcite [2] both only cache the classes.

[1] https://github.com/apache/spark/blob/8f07839e743889bbffe91302f31388ab892bf08a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala#L1463

[2] https://github.com/apache/calcite/blob/a69d24951794f6d3ff67f40b0d117fd0df9a919b/core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableInterpretable.java#L100

Reply:
1. For the execution mode of Flink, the `UserCodeClassLoader` used by each task may be different.
2. The class from CodeGenerated may contain the dependence on user code class, which is in `UserCodeClassLoader` of a task.

So, if we ignore class loader, that will lead to errors. Maybe class not found or class not match.

I don't know the details of Spark and Calcite, but maybe Spark breaks # 2 and Calcite breaks # 1.